Backpack-SwiftUI/Calendar/Classes/Core/CalendarGridCalculator.swift

```diff
@@ -0,0 +1,81 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+protocol CalendarGridCalculator {
+    func calculateCalendarGrid(monthDate: Date) -> [[Date?]]
+}
+
+class InMemoryCacheCalendarGridCalculator: CalendarGridCalculator {
+    private var cache: [Date: [[Date?]]] = [:]
+    private let decoratee: CalendarGridCalculator
+    
+    init(decoratee: CalendarGridCalculator) {
+        self.decoratee = decoratee
+    }
+    
+    func calculateCalendarGrid(monthDate: Date) -> [[Date?]] {
+        if let cached = cache[monthDate] {
+            return cached
+        }
+        let calculated = decoratee.calculateCalendarGrid(monthDate: monthDate)
+        cache[monthDate] = calculated
+        return calculated
+    }
+}
+
+struct DefaultCalendarGridCalculator: CalendarGridCalculator {
+    let calendar: Calendar
+
+    /// Calculated will look like this:
+    ///     [
+    ///         [nil, nil, nil, nil, 1, 2, 3],
+    ///         [4, 5, 6, 7, 8, 9, 10],
+    ///         ...,
+    ///         [27, 28, 29, 30, 31, nil, nil]
+    ///     ]
+    /// Where initial nils are days from previous month, then 1...`numberOfDaysInMonth``,
+    /// then nils to fill the last week
+    func calculateCalendarGrid(monthDate: Date) -> [[Date?]] {
```

Comment:
this logic is a bit complex (as it was before) but now its contained in a structure that can be unit tested! no more guessing work when changing it.